kernel/src/log_segment.rs

```diff
@@ -213,17 +218,144 @@ impl LogSegment {
+        let need_add_actions = checkpoint_read_schema.contains(ADD_NAME);
+        require!(
+            !need_add_actions || checkpoint_read_schema.contains(SIDECAR_NAME),
+        Error::invalid_checkpoint(
+                "If the checkpoint read schema contains file actions, it must contain the sidecar column"
+            )
+        );
+
+        let checkpoint_file_meta: Vec<FileMeta> = checkpoint_parts
             .iter()
             .map(|f| f.location.clone())
             .collect();
-        let checkpoint_stream = engine
-            .get_parquet_handler()
-            .read_parquet_files(&checkpoint_parts, checkpoint_read_schema, meta_predicate)?
-            .map_ok(|batch| (batch, false));
 
-        Ok(commit_stream.chain(checkpoint_stream))
+        let parquet_handler = engine.get_parquet_handler();
+
+        // Historically, we had a shared file reader trait for JSON and Parquet handlers,
+        // but it was removed to avoid unnecessary coupling. This is a concrete case
+        // where it *could* have been useful, but for now, we're keeping them separate.
+        // If similar patterns start appearing elsewhere, we should reconsider that decision.
+        let actions = if checkpoint_parts
+            .first()
+            .is_some_and(|p| p.extension == "json")
+        {
+            engine.get_json_handler().read_json_files(
+                &checkpoint_file_meta,
+                checkpoint_read_schema,
+                meta_predicate,
+            )?
+        } else {
+            parquet_handler.read_parquet_files(
+                &checkpoint_file_meta,
+                checkpoint_read_schema,
+                meta_predicate,
+            )?
+        };
+
+        let actions_iter = actions
+            .map(move |batch_result| -> DeltaResult<_> {
+                let checkpoint_batch = batch_result?;
+                // This closure maps the checkpoint batch to an iterator of batches
+                // by chaining the checkpoint batch with sidecar batches if they exist.
+
+                // 1. In the case where the schema does not contain the add action, we return the checkpoint
+                // batch directly as sidecar files only have to be read when the schema contains the add action.
+                // 2. Multi-part checkpoint batches never have sidecar actions, so the batch is returned as-is.
+                let sidecar_content = if need_add_actions && checkpoint_parts.len() == 1 {
+                    Self::process_sidecars(
+                        parquet_handler.clone(), // cheap Arc clone
+                        log_root.clone(),
+                        checkpoint_batch.as_ref(),
+                    )?
+                } else {
+                    None
+                };
+
+                let combined_batches = std::iter::once(Ok(checkpoint_batch))
+                    .chain(sidecar_content.into_iter().flatten())
+                    .map_ok(|sidecar_batch| (sidecar_batch, false));
```

Comment:
nit: maybe quick comment just for reminder that returning `false` here specifies that it's checkpoint batch